Extend existing case importer validation...

kmtracey opened the pull request:
...to validate dates for fields defined to be dates via data dictionary.

(Intended to illustrate approach of hooking into existing error reporting already in place in case importer.)

## Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Safety Assurance

- [ ] Risk label is set correctly
- [ ] All migrations are backwards compatible and won't block deploy
- [ ] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [ ] If QA is part of the safety story, the "Awaiting QA" label is used
- [ ] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->

### Safety story
<!--
Describe any other pieces to the safety story including
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [ ] This PR can be reverted after deploy with no further considerations

kmtracey committed:
Extend existing case importer validation...

...to validate dates for fields defined to be dates via data dictionary.

(Intended to illustrate approach of hooking into existing error reporting already in place in case importer.)

kmtracey committed:
Add feature flag toggle for case importer data validation.

kmtracey committed:
Merge branch 'master' into case-upload-validate-dates

orangejenny commented on corehq/apps/data_dictionary/models.py:
```diff
@@ -91,3 +94,11 @@ def save(self, *args, **kwargs):
+    def valid_value(self, value):
+        if self.data_type == 'date':
+            try:
+                datetime.strptime(value, '%Y-%m-%d')
```
There's a `ISO_DATE_FORMAT` constant for this in `dimagi.utils.parsing`

orangejenny commented on corehq/apps/data_dictionary/util.py:
```diff
@@ -173,3 +173,13 @@ def get_data_dict_props_by_case_type(domain):
     return set(case_types)
+
+
+def validated_fields(domain, case_type_name):
```
This name sounds like it returns fields that have been validated. How about `fields_to_validate`?

orangejenny commented on corehq/apps/data_dictionary/models.py:
```diff
@@ -91,3 +94,11 @@ def save(self, *args, **kwargs):
         get_data_dict_props_by_case_type.clear(self.case_type.domain)
         return super(CaseProperty, self).save(*args, **kwargs)
+
+    def valid_value(self, value):
```
I'd call this `is_valid` to make it clearer that it's going to return a boolean.

kmtracey replied:
Changed it to `check_validity` since with the movement of importer exception raising into there this ended up not being a method that returns a boolean.

orangejenny commented on corehq/apps/case_importer/do_import.py:
```diff
@@ -250,6 +259,17 @@ def _populate_updated_fields(self, row):
+            if update_field_name in self.validated_fields:
+                case_property = self.validated_fields[update_field_name]
+                if not case_property.valid_value(update_value):
+                    # note: probably want to accumulate full list of invalid
```
Yes, agreed.

kmtracey replied:
![image](https://user-images.githubusercontent.com/430048/125868475-663b87cf-ef49-4333-9f8b-08c5b5537191.png)

Now more than one error will be reported on a row (e.g. row 3 here had multiple columns with date errors).

orangejenny commented on corehq/apps/case_importer/do_import.py:
```diff
@@ -250,6 +259,17 @@ def _populate_updated_fields(self, row):
+                    # full set of info rather than bailing out on first one
+                    # as we're currently doing here.
+                    if case_property.data_type == 'date':
+                        raise exceptions.InvalidDate(update_field_name)
```
What message ultimately gets displayed to the user?

kmtracey replied:
![image](https://user-images.githubusercontent.com/430048/124663312-ea669d00-de77-11eb-8cf5-ccbbdaa22990.png)

This is an example of the message that is eventually given to the user. This `InvalidDate` exception already existed in the code but was apparently unused.

orangejenny replied:
nice, that looks good

orangejenny commented on corehq/apps/case_importer/do_import.py:
```diff
@@ -250,6 +259,17 @@ def _populate_updated_fields(self, row):
+                    # columns for this row and raise an exception with that
+                    # full set of info rather than bailing out on first one
+                    # as we're currently doing here.
+                    if case_property.data_type == 'date':
```
It'd be good to have the data type conditionals in only one place, so that someone adding a new type of validation has to update as few places as possible. This also seems like it could get gnarly if there were multiple kinds of errors for a single data type.

Could this logic go into `valid_value`, which would then probably raise an exception on failure and return None/True on success?

kmtracey replied:
Yeah, my initial intent was to have all the validation details over in the data dictionary side, that got derailed a bit when discovering the existing setup for collecting/reporting errors on import relied on customized exceptions defined in the case importer app. I suppose it would be OK for the data dictionary code to raise the exceptions defined in the case importer app, though it does tie the provider (data dictionary) of the validation more to the "client" using it (case importer) than I would have hoped at the start.

orangejenny replied:
The tighter coupling is a good point. As mentioned the last time we met, I think there are case importer exceptions that may make more sense as part of the data dictionary, but that are in the case importer just because that code is much older. I think the benefit of having all the validation in one place is worth the tighter coupling, and it should be pretty easy to move the exceptions from one app to another in the future.

kmtracey replied:
Ok I went this route of having the data dictionary raise the exception defined in case importer. One wrinkle is that exception wants the field name, which is rather superfluous to send into the validity checker, so it's added by the calling code after the exception is raised. There is other code in do_import that does similar for title and message after creating a generic `CaseRowError` ... not quite the same pattern but similar. Alternatively the column name could be passed to the validity checker solely for the purpose of creating the exception with the right info to begin with. Either works, I'm not sure which would be better.

kmtracey committed:
Address review comments.

(The support for reporting all errors in a row required adding and handling a new exception type that is used to collect and raise together the whole set of exceptions gathered during processing of the row. I can't say I'm too happy with this but it seemed the best way to achieve the desired result without a major restructure of the way these errors/exceptions are currently handled in the importer code. Is there an alternative way I couldn't think of?)

kmtracey committed:
Add some tests of the date validity checking.

orangejenny commented:
@kmtracey Thank you for the updates, I agree this looks good to PR against the dimagi commcare-hq repo.

kmtracey committed:
Merge branch 'master' into case-upload-validate-dates

(Resolved) Conflicts:
	corehq/toggles.py

kmtracey commented:
https://github.com/dimagi/commcare-hq/pull/30095 is the dimagi repo PR for this
